File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
   public void solve(int n,StringBuilder output,List<String>ans,int ind,int open,int close){
       if(ind>=n*2){
           ans.add(output.toString());
           return;
       }
       if(open<n){
          output.append('(');
          solve(n,output,ans,ind+1,open+1,close);
          output.deleteCharAt(output.length()-1);
       }
       if(close<open){
          output.append(')');
          solve(n,output,ans,ind+1,open,close+1);
           output.deleteCharAt(output.length()-1);
       }

   }

    public List<String> generateParenthesis(int n) {

        StringBuilder output=new StringBuilder();
        List<String>ans=new ArrayList<>();
        int open=0;
        int close=0;
        solve(n,output,ans,0,open,close);
        return ans;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void solve(int n,int open,int close,string &output,vector<string>&ans,int ind){
        if(ind>=2*n){
            ans.push_back(output);
            return;
        }
        if(open<n){
            output.push_back('(');
            solve(n,open+1,close,output,ans,ind+1);
            output.pop_back();
        }
        if(close<open){
            output.push_back(')');
            solve(n,open,close+1,output,ans,ind+1);
            output.pop_back();
        }

    }

    vector<string> generateParenthesis(int n) {
        int open=0;
        int close=0;
        string output="";
        vector<string>ans;
        solve(n,open,close,output,ans,0);
        return ans;
    }
};